qutip/solver/integrator.py

```diff
@@ -0,0 +1,262 @@
+    parameters.
+
+    Example
+    -------
+    - The solver cannot reach the desired tolerance within the maximum number
+    of steps.
+    - The step needed to be within desired tolerance is too small.
+    """
+
+
+class Integrator:
+    """
+    A wrapper around ODE solvers.
+    It ensures a common interface for Solver usage.
+    It takes and return states as :class:`qutip.core.data.Data`, it may return
+    a different data-type than the input type.
+
+    Parameters
+    ----------
+    system: qutip.QobjEvo
+        Quantum system in which states evolve.
+
+    options: qutip.SolverOdeOptions
+        Options for the solver.
+
+    Class Attributes
+    ----------------
+    supports_blackbox : bool
+        If True, then the integrator calls only ``system.matmul``,
+        ``system.matmul_data``, ``system.expect``, ``system.expect_data`` and
+        ``isconstant``, ``isoper`` or ``issuper``. This allows the solver using
+        the integrator to modify the system in creative ways. In particular,
+        the solver may modify the system depending on *both* the time ``t``
+        *and* the current ``state`` the system is being applied to.
+
+        If the integrator calls any other methods, set to False.
+
+    supports_time_dependent : bool
+        If True, then the integrator supports time dependent systems. If False,
+        ``supports_blackbox`` should usually be ``False`` too.
+
+    integrator_options : dict
+        A dictionary of options used by the integrator and their default
+        values. Once initiated, ``self.options`` will be a dict with the same
+        keys, not the full options object passed to the solver. Options' keys
+        included here will be supported by the :cls:SolverOdeOptions.
+    """
+    # Used options in qutip.SolverOdeOptions
+    integrator_options = {}
+    # Can evolve time dependent system
+    support_time_dependant = None
+    # Use the QobjEvo's matmul_data method as the driving function
+    supports_blackbox = None
+    name = ""
+
+    def __init__(self, system, options):
+        self.system = system
+        self.options = {
+            **self.integrator_options,
+            **{ key: options[key]
+                for key in self.integrator_options.keys()
+                if key in options and options[key] is not None}
+        }
+        self._is_set = False  # get_state can be used and return a valid state.
+        self._back = (np.inf, None)
+        self._prepare()
+
+    def _prepare(self):
+        """
+        Initialize the solver
+        It should also set the name of the solver to be displayed in Result.
+        """
+        raise NotImplementedError
+
+    def set_state(self, t, state0):
+        """
+        Set the state of the ODE solver.
+
+        Parameters
+        ----------
+        t : float
+            Initial time
+
+        state0 : qutip.Data
+            Initial state.
+
+        .. note:
+            It should set the flags `_is_set` to True.
```

Comment:
Does `set_state` also need to set `self._back` to something now? Perhaps only if the default `mcstep` implementation is used? If yes, we should add a sentence here mentioning what `set_state` should do.

Reply:
No, it's not needed, it's used when the desired time between the 2 previous steps. So having them both at the same values will not help. The value set in `__init__` (`t_back = inf`) is enough to catch bad use (starting the evolution in the wrong direction, which we don't support.)

Reply:
Woot. Let's leave things as they are there.